src/equations/shallow_water_2d.jl

```diff
@@ -0,0 +1,564 @@
+
+Further details for the original finite volume formulation are available in
+- Ulrik S. Fjordholm, Siddhartha Mishr and Eitan Tadmor (2011)
+  Well-balanced and energy stable schemes for the shallow water equations with discontinuous topography
+  [DOI: 10.1016/j.jcp.2011.03.042](https://doi.org/10.1016/j.jcp.2011.03.042)
+and for curvilinear 2D case in the paper:
+- Niklas Wintermeyer, Andrew R. Winters, Gregor J. Gassner and David A. Kopriva (2017)
+  An entropy stable nodal discontinuous Galerkin method for the two dimensional
+  shallow water equations on unstructured curvilinear meshes with discontinuous bathymetry
+  [DOI: 10.1016/j.jcp.2017.03.036](https://doi.org/10.1016/j.jcp.2017.03.036)
+"""
+@inline function flux_nonconservative_fjordholm_etal(u_ll, u_rr,
+                                                     normal_direction_ll::AbstractVector,
+                                                     normal_direction_average::AbstractVector,
+                                                     equations::ShallowWaterEquations2D)
+  # Pull the necessary left and right state information
+  h_ll, _, _, b_ll = u_ll
+  h_rr, _, _, b_rr = u_rr
+
+  # Comes in two parts:
+  #   (i)  Diagonal (consistent) term from the volume flux that uses `normal_direction_average`
+  #        but we use `b_ll` to avoid cross-averaging across a discontinuous bottom topography
+  f2 = normal_direction_average[1] * equations.gravity * h_ll * b_ll
+  f3 = normal_direction_average[2] * equations.gravity * h_ll * b_ll
+
+  #   (ii) True surface part that uses `normal_direction_ll`, `h_average` and `b_jump`
+  #        to handle discontinuous bathymetry
+  h_average = 0.5 * (h_ll + h_rr)
+  b_jump = b_rr - b_ll
+
+  f2 += normal_direction_ll[1] * equations.gravity * h_average * b_jump
+  f3 += normal_direction_ll[2] * equations.gravity * h_average * b_jump
+
+  # First and last equations do not have a nonconservative flux
+  f1 = f4 = zero(eltype(u_ll))
+
+  return SVector(f1, f2, f3, f4)
+end
+
+
+"""
+    flux_fjordholm_etal(u_ll, u_rr, normal_direction,
+                        equations::ShallowWaterEquations2D)
+
+Total energy conservative (mathematical entropy for shallow water equations). When the bottom topography
+is nonzero this should only be used as a surface flux otherwise the scheme will not be well-balanced.
+For well-balancedness in the volume flux use [`flux_wintermeyer_etal`](@ref).
+
+Details are available in Eq. (4.1) in the paper:
+- Ulrik S. Fjordholm, Siddhartha Mishr and Eitan Tadmor (2011)
+  Well-balanced and energy stable schemes for the shallow water equations with discontinuous topography
+  [DOI: 10.1016/j.jcp.2011.03.042](https://doi.org/10.1016/j.jcp.2011.03.042)
+"""
+@inline function flux_fjordholm_etal(u_ll, u_rr, normal_direction::AbstractVector, equations::ShallowWaterEquations2D)
+  # Unpack left and right state
+  h_ll = u_ll[1]
+  v1_ll, v2_ll = velocity(u_ll, equations)
```

Comment:
I went back and forth on this about what would be easiest to work with and how to compute things like the well-balancedness error (but this was before I had the equations struct carry around `H0` as a reference value). Also, for visualization you usually want to plot `H` instead of only `h`. I can change this.

Reply:
We can also keep the current approach, I was just wondering. If you have a good reason for this choice (as you do), please feel free to resolve this.